Desafio3

AquilesAguiar opened the pull request.

AquilesAguiar committed:
Add redis

Adding redis to cache the application

AquilesAguiar committed:
Add logstash

Tirexs integration to the application's logstash, sending the data to elastik search, json handling for ease of use of kibana

ArthurFreitas commented on phoenix/app/cad_products_phoenix/mix.exs:
```diff
@@ -42,7 +42,10 @@ defmodule CadProductsPhoenix.MixProject do
-      {:credo, "~> 1.6", only: [:dev, :test], runtime: false}
+      {:credo, "~> 1.6", only: [:dev, :test], runtime: false},
+      {:redix, "~> 1.1"},
+      {:castore, ">= 0.0.0"},
```
Essa lib está sendo usada?

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/cache.ex:
```diff
@@ -0,0 +1,20 @@
+  def set_product(key, value), do: Redix.command(@conn, ["SET", key, value |> :erlang.term_to_binary |> Base.encode16()])
+
+  # Get products
+  def get_product(key), do: Redix.command(@conn, ["GET", key]) |> decode
```
O style guide recomenda evitar usar o pipe (|>) apenas uma vez. https://github.com/christopheradams/elixir_style_guide#avoid-single-pipelines

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/cache.ex:
```diff
@@ -0,0 +1,20 @@
+  @conn :redis_server
+
+  # Set products
+  def set_product(key, value), do: Redix.command(@conn, ["SET", key, value |> :erlang.term_to_binary |> Base.encode16()])
```
Usar parenteses mesmo em funções unáris no pipe. https://github.com/christopheradams/elixir_style_guide#parentheses-pipe-operator

ArthurFreitas replied:
Pode tirar o sufixo de product dos metodos nesse módulo, ele ta bem genérico e pode aceitar qualquer tipo.

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/cache.ex:
```diff
@@ -0,0 +1,20 @@
+  defp decode({:ok, nil}), do: {:not_found, "key not found"}
+
+  defp decode({:ok, val}) do
+    {:ok, bin} = val |> Base.decode16()
```
Ta difícil a leitura desse metodo, recomendo substituir por variáveis auxiliares ou a estrutura do with. https://elixirschool.com/en/lessons/basics/control_structures#with-3

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/cache.ex:
```diff
@@ -0,0 +1,20 @@
+  @conn :redis_server
+
+  # Set products
+  def set_product(key, value), do: Redix.command(@conn, ["SET", key, value |> :erlang.term_to_binary |> Base.encode16()])
```
No lugar de manter tudo em uma só linha, extrai um método privado que só codifique o produto. Vai facilitar muito a leitura.

ArthurFreitas replied:
Feito tu fez com o decode

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/cache.ex:
```diff
@@ -0,0 +1,20 @@
+  # Get products
+  def get_product(key), do: Redix.command(@conn, ["GET", key]) |> decode
+
+  defp decode({:ok, nil}), do: {:not_found, "key not found"}
```
Nesses casos é mais idiomático usar a tupla {:error, razão}

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -46,4 +58,20 @@ defmodule CadProductsPhoenixWeb.RegisterController do
     end
   end
+
+  defp format_json(products) do
```
Essa função ta com um mal nome porque não formata json e sim indexa produtos no elastic search.

ArthurFreitas replied:
Seria legal extrair esse processo para outro modulo "ProductIndex" que isolasse essa comunicação com o ELS, da mesma forma que tu fez com o cad_products_phoenix/cache.ex

gabrielzuqueto replied:
Ela nem reveria existir na controller. Deveria ficar dentro d eum service.

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -46,4 +58,20 @@ defmodule CadProductsPhoenixWeb.RegisterController do
+      }
+    end)
+
+    Enum.each(products_json, fn(product) -> put("/cad_products/products/#{product.id}", product) end)
```
Um dos nossos usos para o elastic search é desse jeito (indexar produtos para permitir queries de forma rápida sem afetar o DB). Outra forma que utilizamos é para logar eventos para permitir auditoria depois, Ex.: Requests para a API (path, body, http verb, etc), Logs de um job rodando em background, etc.

ArthurFreitas replied:
Porém da forma que ta sendo feita no projeto, esse indice só é atualizado se algum usuário bater no /products. Se tu quiser seguir na proposta de indexar produtos seria legal atualizar o ELS quando o produto é criado/deletado/atualizado.

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -1,6 +1,9 @@
 defmodule CadProductsPhoenixWeb.RegisterController do
```
É bom tratar para que tua aplicação resista a falhas, tanto do redis quanto do elastic search. Tenta rodar ela sem esses serviços e cuida para que ela não quebre nesse caso.

AquilesAguiar committed:
fix cache module

Applying the style guide in the cache module and fixing some functions , creating the product index module for product indexing in els

gabrielzuqueto commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -9,11 +12,19 @@ defmodule CadProductsPhoenixWeb.RegisterController do
   def index(conn, _params) do
-    register = Management.list_register()
-    render(conn, "index.json", register: register)
+    case Cache.get_product("products") do
```
A controller não deve ter lógica. Precisa mover isso para um service.

Outro ponto que é o ideal é usar sempre {:error, "string"} ou {:error, :simbolo}.

Você poderia usar {:error, :not_found}

Vou criar um esboço de service e coloco aqui como exemplo.

gabrielzuqueto replied:
### Exemplo

Criei phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/services/products.ex

```
defmodule CadProductsPhoenix.Services.Products do
  alias CadProductsPhoenix.Cache
  alias CadProductsPhoenix.Management

  def fetch_all do
    case Cache.get_product("products") do
      {:ok, _} = result ->
        result

      _ ->
        register = Management.list_register()
        Cache.set_product("products", register)
        {:ok, register}

    end
  end

  def create(%{"product" => product}) when is_map(product) do
    Cache.delete_product("products")
    Management.create_register(product)
  end

  def create(_params), do: {:error, 422}

  ...
end
```

Daí na controller ficou assim:

```
...
alias CadProductsPhoenix.Services.Products
...
  def index(conn, _params) do
    with {:ok, products} <- Products.fetch_all() do
      render(conn, "index.json", register: products)
    end
  end

  def create(_conn, params) do
    Products.create(params)
  end
```

Em index em caso de erro cai para o FallbackController. Acredito que no create, você colocou 100% no fallback controller (erro e sucesso).

No service retornei {:error, 422} caso não venha o que esperamos, daí no fallback controller, você pode fazer um pattern matching com essa estrutura e daí retorna um 422 padrão.

Isso pode ser reutilizado em toda a aplicação. Assim padroniza o retorno de erro na API.

A controller deve só fazer a ligação com a view. Não deve fazer nada além disso.

gabrielzuqueto commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -9,11 +12,19 @@ defmodule CadProductsPhoenixWeb.RegisterController do
-    render(conn, "index.json", register: register)
+    case Cache.get_product("products") do
+      {:ok, products} ->
+        format_json(products)
```
O retorno do format_json não está sendo utilizado. Vou removê-lo no exemplo

AquilesAguiar committed:
add services

Removing the controller logic and creating a service to aid abstraction

gabrielzuqueto commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/services/product.ex:
```diff
@@ -0,0 +1,48 @@
+      {:ok, _} = result ->
+        Cache.delete("products")
+        ProductIndex.index_products()
+        result
```
Precisa por um default, caso contrário, se retornar erro, vai estourar por falta de match.

```
case ..
  {:ok, ...} ->
    ...

  error ->
     error
end
```

gabrielzuqueto commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/services/product.ex:
```diff
@@ -0,0 +1,48 @@
+      {:ok, product} ->
+        Cache.delete("products")
+        ProductIndex.index_products()
+        {:ok, product}
```
Mesma coisa aqui, precisa por o default, para evitar erro de match

gabrielzuqueto commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/services/product.ex:
```diff
@@ -0,0 +1,48 @@
+
+  def create(_params), do: {:error, 422}
+
+  def update(products, %{"product" => register_params} ) do
```
Como é um item só, ideal é chamar de product. E para ficar mais semântico, no  `{:ok, ...}`, pode por `{:ok, updated_product}`

Ou melhor, no `{:ok, ...}`, como vai retornar esse msm resultado, pode fazer assim:

```
 {:ok, _} = result ->
   Cache.delete("products")
   ProductIndex.index_products()
   result
```

gabrielzuqueto commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/services/product.ex:
```diff
@@ -0,0 +1,48 @@
+
+  def update(_products, _params), do: {:error, 422}
+
+  def delete(products) do
```
Mesmo esquema aqui. Como é um produto só, ideal é alterar para `product`.

gabrielzuqueto commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -5,55 +5,34 @@ defmodule CadProductsPhoenixWeb.RegisterController do
   alias CadProductsPhoenix.Management
```
Esqueceu de remover os alias que não utiliza mais.

gabrielzuqueto commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/fallback_controller.ex:
```diff
@@ -27,4 +31,11 @@ defmodule CadProductsPhoenixWeb.FallbackController do
     |> render(:"404")
   end
+
+  def call(conn, {:error, 422}) do
```
Está correto, porém, como vai exibir sempre o código do erro, pode deixar mais genérico. Mas está correto da forma que fez. É só um exemplo a mais.

```
  def call(conn, {:error, error_code}) when is_number(error_code) do
    conn
    |> put_status(error_code)
    |> put_view(CadProductsPhoenixWeb.ErrorView)
    |> render("#{error_code}")
  end
```

Ah... em um projeto real, o ideal é descrever melhor  o erro. Para isso poderia fazer algo como:

```
  def call(conn, {:error, %{code: error_code, message: message}}) when is_number(error_code) do
    conn
    |> put_status(error_code)
    |> put_view(CadProductsPhoenixWeb.ErrorView)
    |> render(message)
  end
  ```
Daí pode manipular os erros para retornarem assim:

```
{:error, %{code: 422, message: "Mensagem que ajude o cliente a corrigir o erro"}}
```

Como falei, é só mais um exemplo msm.

gabrielzuqueto commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/fallback_controller.ex:
```diff
@@ -27,4 +31,11 @@ defmodule CadProductsPhoenixWeb.FallbackController do
+    conn
+    |> put_status(:unprocessable_entity)
+    |> put_view(CadProductsPhoenixWeb.ErrorView)
+    |> render(:"422")
```
Não precisa criar um symbol, poderia por direto `render(422)`.

AquilesAguiar committed:
fix fallback error

Generalizing the error fallback so that other codes can be used, removing unused aliases and adding the default in the case

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/fallback_controller.ex:
```diff
@@ -25,6 +29,20 @@ defmodule CadProductsPhoenixWeb.FallbackController do
+    conn
+    |> put_status(error_code)
+    |> put_view(CadProductsPhoenixWeb.ErrorView)
+    |>  send_resp(error_code, message)
```
Sobrou espaço

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/fallback_controller.ex:
```diff
@@ -25,6 +29,20 @@ defmodule CadProductsPhoenixWeb.FallbackController do
+    |> render(404)
   end
+
+  # def call(conn, {:error, 422}) do
```
Sempre remova código comentado

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -46,4 +46,5 @@ defmodule CadProductsPhoenixWeb.RegisterController do
       |> halt()
     end
   end
+
```
Sobrou espaço

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/services/product.ex:
```diff
@@ -0,0 +1,51 @@
+  def fetch_all() do
+    case Cache.get("products") do
+      {:ok, _} = result ->
+        ProductIndex.index_products()
```
Dessa forma vai pesar muito tua aplicação, sempre que o cliente acessar o index terá que buscar todos os produtos do banco e indexa-los 1 a 1 no ELS. Uma forma melhor seria: indexar o produto quando for criado/atualizado e removê-lo do indíce quando for deletado.

AquilesAguiar committed:
Fix els indexing

Changing the way elsk indexes, instead of redoing the query in the database, it only indexes when it is created/updated

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/product_index.ex:
```diff
@@ -0,0 +1,25 @@
+
+  def index_product(product) do
+    {:ok, prod} = product
+    IO.inspect(prod)
```
Sobrou o inspect

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/product_index.ex:
```diff
@@ -0,0 +1,25 @@
+  alias CadProductsPhoenix.Management
+
+  def index_product(product) do
+    {:ok, prod} = product
```
Faz esse matching na definição da função.

`def index_product({:ok, prod})`

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/services/product.ex:
```diff
@@ -0,0 +1,51 @@
+
+  def update(product, %{"product" => register_params} ) do
+    case Management.update_register(product, register_params) do
+      {:ok, _} ->
```
Faltou pegar o produto atualizado aqui não? Tu ta repassando o product velho para o resto da função.

AquilesAguiar committed:
fix return update

retuning the update product

AquilesAguiar committed:
fix the match update

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/product_index.ex:
```diff
@@ -0,0 +1,23 @@
+
+  def index_product({:ok, prod}) do
+    product_json =
+       %{
```
É bom ter um dado temporal:

`last_update_at: DateTime.to_iso8601(DateTime.utc_now())`

AquilesAguiar committed:
add last_update
